examples/jsm/loaders/TaskManager.js

```diff
@@ -0,0 +1,560 @@
+/**
+ * @author Don McCurdy / https://www.donmccurdy.com
+ * @author Kai Salmen / https://kaisalmen.de
+ */
+
+import { FileLoaderBufferAsync } from "./obj2/utils/FileLoaderBufferAsync.js";
+
+/**
+ *
+ */
+class TaskManager {
+
+    constructor () {
+
+        this.types = new Map();
+        this.verbose = false;
+
+    }
+
+    /**
+     * Set if logging should be verbose
+     *
+     * @param {boolean} verbose
+     * @return {TaskManager}
+     */
+    setVerbose ( verbose ) {
+
+        this.verbose = verbose;
+        return this;
+
+    }
+
+    /**
+     * Returns true if support for the given task type is available.
+     * @param {string} type The type as string
+     * @return boolean
+     */
+    supportsType ( type ) {
+
+        return this.types.has( type );
+
+    }
+
+    /**
+     * Registers functionality for a new task type.
+     * @param {string} type The name to be used for registration.
+     * @param {number} maximumWorkerCount How many workers are allows. Set 0 for execution on Main ({@link FakeTaskWorker})
+     * @param {function} initFunction The function to be called when the worker is initialised
+     * @param {function} executeFunction The function to be called when the worker is executed
+     * @param {function} comRoutingFunction The function that should handle communication, leave undefined for default behavior
+     * @param {String[]} [dependencyUrls]
+     * @return {TaskManager}
+     */
+    registerType ( type, maximumWorkerCount, initFunction, executeFunction, comRoutingFunction, dependencyUrls ) {
```

Comment:
Are there benefits to having a separate maxWorkerCount for each type of task? The downside, I think, is that you can't effectively load balance across different types of task. If any worker can do any task, that balancing could be easier.

Reply:
Maybe there is need for an overall count and a desired per `taskType` count. Was that your idea behind cost parameter?

> If any worker can do any task, that balancing could be easier.

Is this a realistic scenario?

Reply:
I consider the `cost` parameter unrelated to whether or not there are separate worker queues for different task types.

My original idea had been to have a configurable number of workers for the entire TaskManager, say N. Each worker can perform any type of task, and all tasks are balanced across those N workers. So there would be no per-task worker limit, just the TaskManager total limit.

> Is this a realistic scenario?

Having multiple tasks running in parallel is realistic, at least. For example, when using both compressed geometry and compressed textures you'll need DRACOLoader and (one of) KTX2Loader or BasisTextureLoader. 

Whether there are very many realistic situations where shared queues are likely to give better performance? I'm not sure. If you think this is more complicated, or worse for some other reason, I'm also fine with leaving this as you have it for now.

Reply:
> My original idea had been to have a configurable number of workers for the entire TaskManager, say N. Each worker can perform any type of task, and all tasks are balanced across those N workers. So there would be no per-task worker limit, just the TaskManager total limit.

Ok, now I see your point. Then we need to create N workers for every taskType either lazy or on init (cost are only extra memory and init time what becomes negligible if Workers really do intensive work). Then we can always use maximum of N for every taskType.
From my point of view the Worker should only include the code domain (init, exec) of one task as otherwise I fear mixing task and their dependencies into the worker will introduce unnecessary complexity and mixing module and non-module code will not be manageable. I hope this makes sense.

Reply:
> Then we need to create N workers for every taskType... Then we can always use maximum of N for every taskType depending on current demand.

This is a bit different than what I meant. Something like...

```
var manager = new TaskManager().setWorkerLimit(2);
var objLoader = new OBJLoader(manager);
var ktx2Loader = new KTX2Loader(manager);
```

... would create only two Web Workers *total*, no matter how many loaders are using that manager, and no matter how many resources are loaded. OBJ and KTX2 tasks could be balanced across those two workers, and a single worker's queue would have tasks of both types.

There are some very real advantages to that approach: you have firm control over the _total_ number of Web Workers, and work is evenly distributed. But the concern of mixing module and non-module code makes sense. I'm not sure the module worker type is going to be worth a whole separate API in the long term, but I don't want to hold up this PR on the question – it's fine with me to continue as you have the PR for now.

Reply:
> There are some very real advantages to that approach: you have firm control over the total number of Web Workers, and work is evenly distributed.

I agree and I like the simplicity of your approach. Unfortunately modules and dependencies make it very complicated to combine all code in one worker and then have n instances of them. I hope my understanding matches what you meant now.

What about the following compromise:
We follow the strict worker limit. The proposed `TaskManager` could be modified to internally create N worker for each registered loader. Taking your example this means 2*OBJLoader2 workers + 2*KTX2Loader workers are created. When tasks are add/executed it only runs two in parallel (2*OBJ+0*KTX, 1*OBJ+1*KTX, 0*OBJ + 2*KTX), the other spares are not utilized. Worker creation should then move to the *register* step. A *dispose* function could be added to the worker contract to signal runtime data clean-up.

Reply:
> When tasks are add/executed it only runs two in parallel ...

This will be very complicated to enforce. What happens if we get requests in this order:

1. 10 OBJ parse tasks
2. 2 KTX2 parse tasks

After step (1) I would expect that the manager assigns ~5 OBJ tasks to each OBJ worker. But then when it gets to step (2) there are already two workers running, do we have to wait until one of the workers clears its queue before assigning KTX tasks to a KTX worker? What if more OBJ tasks arrive in the meantime?

I would prefer to avoid all that... maybe we have a `setWorkerLimit( n )` method on the TaskManager, and for now it will create N workers _per task type_ and use them all. Then in the future, if we decide to, we can try to combine the worker pool so that tasks share workers for more equitable load balancing.

Reply:
Add task puts the processing requests to a fifo queue (currently Promises are stored in array). Whenever a worker signals exec complete a task is taken from the queue and a fitting worker is kicked. What is done in the current `TaskManager`/example is not so far off expect for the overall count is per task type and not globally. But getting there not be hard. I will prototype this.

Reply:
I see! I wonder whether the time to signal the main thread that a task is done and then wait for a new task to come back has a measurable impact on completion times? Thinking of a model with 100+ Draco geometries for example... that doesn't seem like much of a problem though.

Reply:
In lastest commit the maximum amount (N) of parallel worker execution is now handled on `TaskManager` level. N workers are created of each type (except for main fallback), but only the maximum defined in `TaskManager` are executed in parallel.

Reply:
The previous code did not work as described. I pushed another update that really does works the way it does.
Init time of workers increases linearly with quantity even if Blobs/Strings are cached. For module workers there seems no obvious way to cache code except for pre-loading the module in the embedding html.

Execution on main is still strange. The `FakeTaskWorker` is now treated identically during execution, but when it solely used in the example than it completely blocks any rendering. When it is intermixed with real workers, then the rendering is not blocked. Manual triggering of requestAnimationFrame did not help. For now I leave it as it is